cpp/tests/experimental/spatial/linestring_intersection_test.cu

```diff
@@ -0,0 +1,447 @@
+
+  auto multilinestrings1 = make_multilinestring_array({0, 1}, {0, 2}, {P{0.0, 0.0}, P{1.0, 1.0}});
+
+  auto multilinestrings2 =
+    make_multilinestring_array({0, 1}, {0, 4}, {P{0.0, 1.0}, P{1.0, 0.0}, P{0.5, 0.0}, P{0.5, 1}});
+
+  auto expected = make_linestring_intersection_result<T, index_t, types_t>(
+    {0, 1}, {0}, {0}, {P{0.5, 0.5}}, {}, {0}, {0}, {0}, {0}, this->stream(), this->mr());
+
+  CUSPATIAL_RUN_TEST(this->template run_single_test<index_t>,
+                     multilinestrings1.range(),
+                     multilinestrings2.range(),
+                     expected);
+}
+
+TYPED_TEST(LinestringIntersectionTest, SingletoSingleOnePairWithMergeableSegment)
+{
+  using T = TypeParam;
+  using P = vec_2d<T>;
+
+  using index_t = typename linestring_intersection_result<T, std::size_t>::index_t;
+  using types_t = typename linestring_intersection_result<T, std::size_t>::types_t;
+
+  auto multilinestrings1 = make_multilinestring_array({0, 1}, {0, 2}, {P{0.0, 0.0}, P{3.0, 3.0}});
+
+  auto multilinestrings2 = make_multilinestring_array(
+    {0, 1}, {0, 4}, {P{0.0, 0.0}, P{1.0, 1.0}, P{0.5, 0.5}, P{1.5, 1.5}});
+
+  auto expected =
+    make_linestring_intersection_result<T, index_t, types_t>({0, 1},
+                                                             {1},
+                                                             {0},
+                                                             {},
+                                                             {segment<T>{P{0.0, 0.0}, P{1.5, 1.5}}},
+                                                             {0},
+                                                             {0},
+                                                             {0},
+                                                             {0},
+                                                             this->stream(),
+                                                             this->mr());
+
+  CUSPATIAL_RUN_TEST(this->template run_single_test<index_t>,
+                     multilinestrings1.range(),
+                     multilinestrings2.range(),
+                     expected);
+}
+
+TYPED_TEST(LinestringIntersectionTest, SingletoSingleOnePairWithMergeablePoint)
+{
+  using T = TypeParam;
+  using P = vec_2d<T>;
+
+  using index_t = typename linestring_intersection_result<T, std::size_t>::index_t;
+  using types_t = typename linestring_intersection_result<T, std::size_t>::types_t;
+
+  auto multilinestrings1 = make_multilinestring_array({0, 1}, {0, 2}, {P{0.0, 0.0}, P{3.0, 3.0}});
+
+  auto multilinestrings2 = make_multilinestring_array(
+    {0, 1}, {0, 4}, {P{0.0, 1.0}, P{1.0, 0.0}, P{2.0, 2.0}, P{0.0, 0.0}});
+
+  auto expected =
+    make_linestring_intersection_result<T, index_t, types_t>({0, 1},
+                                                             {1},
+                                                             {0},
```

Comment:
A similar problem exists here where the user cannot recover which points in `rhs` intersect.

Reply:
#860